File: qn_go/qn_go_en/qn_en.go
package main

import (
	"fmt"
	"math"
)

func main() {
	// made by sm300max
	fmt.Println("made by sm300max")

	var a, b, c float64
	var is string

	fmt.Println("Enter a")
	fmt.Scan(&a)

	fmt.Println("Enter b:")
	fmt.Scan(&b)

	fmt.Println("Enter c:")
	fmt.Scan(&c)

	fmt.Println("Choose [</<=/=/=>/>]:")
	fmt.Scan(&is)

	if a == 0 {
		fmt.Println("A square expression cannot contain a coefficient a equal to 0")
		return
	}

	if is != "<" && is != "<=" && is != "=" && is != "=>" && is != ">" {
		fmt.Println("A mistake was made in choosing the sign, try again.")
		return
	}

	var q string

	fmt.Println(fmt.Sprint(a) + "x²" + "+" + fmt.Sprint(b) + "x" + "+" + fmt.Sprint(c) + is + "0")
	fmt.Println("The expression is correct? [Y/n]")
	fmt.Scanln(&q)

	if q != "y" && q != ""{
		fmt.Println("Try again")
		return
	}
	var ud int8
	fmt.Println("In view of the function:")
	fmt.Println("y=" + fmt.Sprint(a) + "x²" + "+" + fmt.Sprint(b) + "x" + "+" + fmt.Sprint(c))
	fmt.Println("The square function")
	fmt.Println("The graph is a parabola")
	if a > 0 {
		ud = 1
		fmt.Println("Branches up")
	} else {
		ud = 2
		fmt.Println("Branches down")
	}

	fmt.Println("Function zeros: " + fmt.Sprint(a) + "x²" + "+" + fmt.Sprint(b) + "x" + "+" + fmt.Sprint(c) + "=0")

	var x1, x2 float64
	var x float64
	var oot int8

	D := b*b - 4*a*c
	if D < 0 {
		fmt.Println("D = b² - 4ac")
		fmt.Println("D = " + fmt.Sprint(b) + "² - 4*" + fmt.Sprint(a) + "*" + fmt.Sprint(c))
		fmt.Println("D = " + fmt.Sprint(D))
		fmt.Println("D < 0")
		fmt.Println("The expression has no roots")
		fmt.Println("It's decided")
	} else if D == 0 {
		oot = 1
		x = -b / (2 * a)
		fmt.Println("D = b² - 4ac")
		fmt.Println("D = " + fmt.Sprint(b) + "² - 4*" + fmt.Sprint(a) + "*" + fmt.Sprint(c))
		fmt.Println("D = " + fmt.Sprint(D))
		fmt.Println("D = 0")

		fmt.Println("The expression has one root")
		fmt.Println("x = -b / 2a")
		fmt.Println("x = -" + fmt.Sprint(b) + "/2" + fmt.Sprint(a))
		fmt.Println("x = " + fmt.Sprint(x))
		fmt.Println("It's decided")
	} else if D > 0 {
		oot = 2
		x1 = (-b + math.Sqrt(D)) / (2 * a)
		x2 = (-b - math.Sqrt(D)) / (2 * a)
		fmt.Println("D = b² - 4ac")
		fmt.Println("D = " + fmt.Sprint(b) + "² - 4*" + fmt.Sprint(a) + "*" + fmt.Sprint(c))
		fmt.Println("D = " + fmt.Sprint(D))
		fmt.Println("D > 0")
		fmt.Println("The expression has two roots")
		fmt.Println("x = (-b ± √D) / (2a)")

		if a < 0 && b < 0 {
			fmt.Println("(x1=-(" + fmt.Sprint(b) + ")+√" + fmt.Sprint(D) + ") / (2*(" + fmt.Sprint(a) + "))")
			fmt.Println("(x1=-(" + fmt.Sprint(b) + ")-√" + fmt.Sprint(D) + ") / (2*(" + fmt.Sprint(a) + "))")
		} else if b < 0 {
			fmt.Println("(x1=-(" + fmt.Sprint(b) + ")+√" + fmt.Sprint(D) + ") / (2*" + fmt.Sprint(a) + ")")
			fmt.Println("(x1=-(" + fmt.Sprint(b) + ")-√" + fmt.Sprint(D) + ") / (2*" + fmt.Sprint(a) + ")")
		} else if a < 0 {
			fmt.Println("(x1=-" + fmt.Sprint(b) + "+√" + fmt.Sprint(D) + ") / (2*(" + fmt.Sprint(a) + "))")
			fmt.Println("(x1=-" + fmt.Sprint(b) + "-√" + fmt.Sprint(D) + ") / (2*(" + fmt.Sprint(a) + "))")
		} else {
			fmt.Println("(x1=-" + fmt.Sprint(b) + "+√" + fmt.Sprint(D) + ") / (2*" + fmt.Sprint(a) + ")")
			fmt.Println("(x1=-" + fmt.Sprint(b) + "-√" + fmt.Sprint(D) + ") / (2*" + fmt.Sprint(a) + ")")
		}
		fmt.Println("x1 = " + fmt.Sprint(x1))
		fmt.Println("x2 = " + fmt.Sprint(x2))
	}

	var start, end float64
	switch {
	case x1 > x2:
		start = x2
		end = x1
	case x1 < x2:
		start = x1
		end = x2
	}

	if oot == 2 {

		if ud == 1 {
			if is == ">" || is == "=>" {
				if is == ">" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(start) + ")∪(" + fmt.Sprint(end) + ";+∞)")
				} else if is == "=>" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(start) + "]∪[" + fmt.Sprint(end) + ";+∞)")
				}
			} else if is == "<" || is == "<=" {
				if is == "<" {
					fmt.Println("Answer: (" + fmt.Sprint(start) + ";" + fmt.Sprint(end) + ")")
				} else if is == "<=" {
					fmt.Println("Answer: [" + fmt.Sprint(start) + ";" + fmt.Sprint(end) + "]")
				}
			}
		} else if ud == 2 {
			if is == ">" || is == "=>" {
				if is == ">" {
					fmt.Println("Answer: (" + fmt.Sprint(start) + ";" + fmt.Sprint(end) + ")")
				} else if is == "=>" {
					fmt.Println("Answer: [" + fmt.Sprint(start) + ";" + fmt.Sprint(end) + "]")
				}
			} else if is == "<" || is == "<=" {
				if is == "<" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(start) + ")∪(" + fmt.Sprint(end) + ";+∞)")
				} else if is == "<=" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(start) + "]∪[" + fmt.Sprint(end) + ";+∞)")
				}
			}
		}
	} else if oot == 1 {
		if ud == 1 {

			if is == "=>" || is == "<=" {
				if is == "=>" {
					fmt.Println("Answer: (-∞;+∞)")
				} else if is == "<=" {
					fmt.Println("Answer: There are no solutions(⌀)")
				}
			} else if is == ">" || is == "<" {
				if is == ">" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(x) + ")∪(" + fmt.Sprint(x) + ";+∞)")
				} else if is == "<" {
					fmt.Println("Answer: There are no solutions(⌀)")
				}
			}
		} else if ud == 2 {
			if is == "=>" || is == "<=" {
				if is == "=>" {
					fmt.Println("Answer: There are no solutions(⌀)")
				} else if is == "<=" {
					fmt.Println("Answer: (-∞;+∞)")
				}
			} else if is == ">" || is == "<" {
				if is == ">" {
					fmt.Println("Answer: There are no solutions(⌀)")
				} else if is == "<" {
					fmt.Println("Answer: (-∞;" + fmt.Sprint(x) + ")∪(" + fmt.Sprint(x) + ";+∞)")
				}
			}
		}
	}
	var wait string
	fmt.Scan(&wait)
}

File: qn_rust/qn_rust_en/qn_en.rs
use std::io;

fn main() {
    let mut antmf = String::new();

    println!("Enter the first coefficient:");

    io::stdin()
        .read_line(&mut antmf)
        .expect("error");

    let a: f64 = antmf.trim().parse().expect("Please enter a valid number");

    if a == 0.0 {
        println!("Quadratic expression cannot have a equal to zero");
        return;
    }

    let mut bntmf = String::new();

    println!("Enter the second coefficient:");

    io::stdin()
        .read_line(&mut bntmf)
        .expect("error");

    let b: f64 = bntmf.trim().parse().expect("Please enter a valid number");

    let mut cntmf = String::new();

    println!("Enter the third coefficient:");

    io::stdin()
        .read_line(&mut cntmf)
        .expect("error");

    let c: f64 = cntmf.trim().parse().expect("Please enter a valid number");

    let mut ismoded = String::new();

    println!("Enter operation sign:");

    io::stdin()
        .read_line(&mut ismoded)
        .expect("error");
    
    let is = ismoded.trim();

    if is != "<" && is != "<=" && is != "=" && is != "=>" && is != ">" {
        println!("Error in sign selection, please try again");
        return;
    }

    let mut question = String::new();

    println!("Is the expression correct?");
    println!("{}x² + {}x + {} {} 0 [y/n]", a, b, c, is);

    io::stdin()
        .read_line(&mut question)
        .expect("error");

    let q = question.trim();

    if q != "y" {
        println!("Please try again");
        println!("You entered {}", q);
        return;
    }

    let upordown: u8;
    if a > 0.0 {
        upordown = 1;
    } else if a < 0.0 {
        upordown = 2;
    } else {
        upordown = 0;
    }
    
    println!("Function: y = {}x² + {}x + {}", a, b, c);
    println!("Quadratic function");
    println!("Graph is a parabola");

    match upordown {
        1 => println!("Branches upward"),
        2 => println!("Branches downward"),
        _ => println!("error")
    }
    println!("Function zeros: {}x² + {}x + {} = 0", a, b, c);

    let oneortwo: u8;
    let x: f64;
    let mut xs: (f64, f64) = (0.0, 0.0);
    let d: f64;
    d = (b*b) - (4.0*a*c);

    if d < 0.0 {
        xs.0 = 0.0;
        xs.1 = 0.0;
        x = 0.0;
        oneortwo = 0;
        println!("D = b² - 4ac");
        println!("D = {}² - 4 * {} * {}", b, a, c);
        println!("D = {}", d);
        println!("D < 0");
        println!("The expression has no roots");
        println!("Solved");
    } else if d == 0.0 {
        oneortwo = 1;
        x = -b / (2.0 * a);
        xs.0 = 0.0;
        xs.1 = 0.0;
        println!("D = b² - 4ac");
        println!("D = {}² - 4 * {} * {}", b, a, c);
        println!("D = {}", d);
        println!("The expression has one root");
        println!("x = -b / 2a");
        println!("x = -{} / 2{}", b, a);
        println!("x = {}", x);
        println!("Solved");
    } else if d > 0.0 {
        oneortwo = 2;
        xs.0 = (-b + d.sqrt())/(2.0*a);
        xs.1 = (-b - d.sqrt())/(2.0*a);
        x = 0.0;
        println!("D = b² - 4ac");
        println!("D = {}² - 4 * {} * {}", b, a, c);
        println!("D = {}", d);
        println!("The expression has two roots");
        println!("x = (-b ± √D) / (2a)");

        if a < 0.0 && b < 0.0 {
            println!("x1 = (-({}) + √{}) / 2 * ({}) )", b, d, a);
            println!("x2 = (-({}) - √{}) / 2 * ({}) )", b, d, a);
        } else if b < 0.0 {
            println!("x1 = (-({}) + √{}) / 2 * {} )", b, d, a);
            println!("x2 = (-({}) - √{}) / 2 * {} )", b, d, a);
        } else if a < 0.0 {
            println!("x1 = ( -{} + √{}) / 2 * ({}) )", b, d, a);
            println!("x2 = ( -{} - √{}) / 2 * ({}) )", b, d, a);
        } else {
            println!("x1 = (-{} + √{}) / 2 * {} )", b, d, a);
            println!("x2 = (-({}) - √{}) / 2 * {} )", b, d, a);
        }
        println!("x1 = {}", xs.0);
        println!("x2 = {}", xs.1);
    } else {
        x = 0.0;
        oneortwo = 0;
    }

    let start: f64;
    let end: f64;

    if xs.0 < xs.1 {
        start = xs.0;
        end = xs.1;
    } else if xs.0 > xs.1 {
        start = xs.1;
        end = xs.0;
    } else {
        start = 0.0;
        end = 0.0;
    }

    if oneortwo == 2 {
        if upordown == 1 {
            if is == ">" || is == "=>" {
                if is == ">" {
                    println!("Answer: (-∞;{})∪({};+∞)", start, end);
                } else if is == "=>" {
                    println!("Answer: (-∞;{}]∪[{};+∞)", start, end);
                }
            }
            if is == "<" || is == "<=" {
                if is == "<" {
                    println!("Answer: ({};{})", start, end);
                }
                if is == "<=" {
                    println!("Answer: [{};{}]", start, end);
                }
            }
        }
        if upordown == 2 {
            if is == ">" || is == "=>" {
                if is == ">" {
                    println!("Answer: ({};{})", start, end)
                }
                if is == "=>" {
                    println!("Answer: [{};{}]", start, end);
                }
            }
            if is == "<" || is == "<=" {
                if is == "<" {
                    println!("Answer: (-∞;{})∪({};+∞)", start, end);
                } else if is == "<=" {
                    println!("Answer: (-∞;{}]∪[{};+∞)", start, end);  
                }               
            }
        }
    } else if oneortwo == 1 {
        if upordown == 1 {
            if is == "=>" || is == "<=" {
                if is == "=>" {
                    println!("Answer: (-∞;+∞)");
                } else if is == "<=" {
                    println!("Answer: No solution (∅)");
                }
            } else if is == ">" || is == "<" {
                if is == ">" {
                    println!("Answer: (-∞;{})∪({};+∞)", x, x);
                } else if is == "<" {
                    println!("Answer: No solution (∅)");
                }
            }
        } else if upordown == 2 {
            if is == "=>" || is == "<=" {
                if is == "=>" {
                    println!("Answer: No solution (∅)");
                } else if is == "<=" {
                    println!("Answer: (-∞;+∞)");
                }
            } else if is == ">" || is == "<" {
                if is == ">" {
                    println!("Answer: No solution (∅)");
                } else if is == "<" {
                    println!("Answer: (-∞;{})∪({};+∞)", x, x);
                }
            }
        }
    }

    let mut exitwaiting = String::new();
    io::stdin()
        .read_line(&mut exitwaiting)
        .expect("error");

    // made by sm300max
}